pkg/controller/controller_binding.go

```diff
@@ -366,12 +379,16 @@ func (c *controller) injectBinding(binding *v1alpha1.Binding, credentials map[st
 
 	found := false
 
-	_, err := c.kubeClient.Core().Secrets(binding.Namespace).Get(binding.Spec.SecretName, metav1.GetOptions{})
+	existingSecret, err := c.kubeClient.Core().Secrets(binding.Namespace).Get(binding.Spec.SecretName, metav1.GetOptions{})
 	if err == nil {
 		found = true
 	}
 
 	if found {
+		if !isOwnedByBinding(&existingSecret.ObjectMeta, binding) {
```

Comment:
I think that the proper way of handling both terminal and temporary errors is to perform cleanup in service broker at the beginning of processing as part of reconciliation loop. And instead of performing it right there (as it could fail as well, the Service Catalog server could be restarted, and so on), we should just mark it with error condition. The example of conditions we might have:
- needs cleanup: true/false
- terminal error: true/false

I'll think more about implementing such pattern to solve it once for every non-success case.